### ISSUE TITLE
How to extract initial batch nodes from `DataFlow`

### ISSUE THREAD

**Issue by duncster94** [rusty1s/pytorch_geometric#498]
## ❓ Questions & Help

I'm trying to subset node features from the output of a layer using `DataFlow` with the initial nodes in the batch.

For example, I have a `DataFlow` object like `(3<-5<-7<-9)`. When I use `GATConv` with this `DataFlow`, I will get the feature tensors `x1, x2, x3` where `x1.shape = (7, f1)`, `x2.shape = (5, f2)` and `x3.shape = (3, f3)` where `f1, f2, f3` are the output sizes of each layer (not important here). I'll call the 3 nodes in `x3` the `batch_nodes`. I want to be able to subset the features from both `x1` and `x2` using `batch_nodes`. The problem here is that the node indices change with each hop of the `DataFlow` so I can't keep track of which indices in the feature tensor correspond to `batch_nodes`. Any ideas? 

For context, I'm trying to add residuals from each GAT layer to the output of the final GAT layer which is why I need the `batch_nodes` features from `x1`, `x2` (i.e. so I can do `x = x1[batch_nodes] + x2[batch_nodes] + x3`).

**Comment by rusty1s** [comment]
Working on improving the `NeighborSampler` :)

**Comment by duncster94** [comment]
Awesome, thanks again for all your help.

**Comment by rusty1s** [comment]
Hi, I made some changes to the `NeighborSampler`:

1. In case `add_self_loops=True`, each `Block` element in the `DataFlow` object now comes with a `res_n_id` attribute to select the nodes from the previous layer (not just the "batch" nodes) for implementing skip connections.
2. I added a `bipartite=False` option to the `NeighborSampler`, which, instead of building a computation flow from neighboring nodes to batch nodes, is building a complete subgraph. Instead of a `DataFlow` object, it returns a `Data` object that holds the subgraph in `edge_index`, a map from subgraph nodes to original node identifiers in `n_id`, a map from batch nodes to original node identifiers in `b_id`, and a map from batch nodes to the subgraph node identifiers in `sub_b_id`.

This may get more clearer when looking at its [test suite](https://github.com/rusty1s/pytorch_geometric/blob/master/test/data/test_sampler.py#L50-L95).

**Comment by duncster94** [comment]
This is perfect! I've decided to keep `bipartite=True` and use `res_n_id` to subset the feature matrices since it seems to be a bit faster than option 2.

Final question, is it possible for `NeighborSampler` to also handle edge attributes? For example, if I had a weighted graph I would want both `edge_index` and `values` from `DataFlow` where `edge_index.shape[1] == values.shape[0]`. Then I could easily modify the `softmax` in `GATConv` to handle weights.

**Comment by rusty1s** [comment]
Tricky question! We provide an `e_id` attribute for filtering edge features, but we remove it in case `add_self_loops` is added as an argument (where one creates an internal self-loop for which no edge weight exists). Will work on it :)

**Comment by rusty1s** [comment]
So, I just pushed an updated version to master. Each block element now has an `e_id` attribute which maps each edge in the bipartite graph to the original edge of your data. You can now easily filter your edge features via `data.edge_attr[block.e_id]` and pass it to the respective convolution operator.

It gets trickier if the `add_self_loops` argument is set to `True`. In this case, PyG adds remaining self-loops to the graph (for which no edge features exist). Those edges are marked by a `-1` in the `e_id` vector and you need to be aware to manually set them, *e.g.*:
```python
edge_attr = data.edge_attr[block.e_id]`
edge_attr[block.e_id == -1] = 1
```
This is not that cool, but you can prevent this easily by adding self-loops to your graph *before* passing in your data object to the `NeighborSampler`. A better way may be to force this behaviour: If one wants to access `e_id` with `add_self_loops=True`, one has to manually pre-add self-loops to the graph. WDYT?

**Comment by duncster94** [comment]
Fantastic, it's working beautifully. Since I've opted to use `DataFlow` (i.e. `bipartite = True`), I set `add_self_loops = True` and do the manual indexing. My graphs have self-loops added manually so I can avoid doing `edge_attr[block.e_id == -1] = 1`.

**Comment by rusty1s** [comment]
Perfect :)
